privilege/privileges/tidb_auth_token.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package privileges
+
+import (
+	"context"
+	"encoding/json"
+	"strings"
+	"sync"
+	"time"
+
+	jwkRepo "github.com/lestrrat-go/jwx/v2/jwk"
+	jwsRepo "github.com/lestrrat-go/jwx/v2/jws"
+	jwtRepo "github.com/lestrrat-go/jwx/v2/jwt"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+// JWKSImpl contains a JSON Web Key Set (JWKS), and a filepath that stores the JWKS
+type JWKSImpl struct {
+	// contain one more JWKS for recover when failing to load the JWKS from the filepath.
+	cur, last jwkRepo.Set
+	m         sync.RWMutex
+	filepath  string
+}
+
+// GlobalJWKS is the global JWKS for tidb-server
+var GlobalJWKS JWKSImpl
+
+func (jwks *JWKSImpl) load() (err error) {
+	jwks.m.Lock()
+	defer jwks.m.Unlock()
+	jwks.last = jwks.cur
+	jwks.cur, err = jwkRepo.ReadFile(jwks.filepath)
+	if err != nil {
+		jwks.cur = jwks.last
+	}
+	return err
+}
+
+func (jwks *JWKSImpl) verify(tokenBytes []byte) (payload []byte, err error) {
+	jwks.m.RLock()
+	defer jwks.m.RUnlock()
+	return jwsRepo.Verify(tokenBytes, jwsRepo.WithKeySet(jwks.cur))
+}
+
+// LoadJWKS4AuthToken reload the jwks every auth-token-refresh-interval.
+func (jwks *JWKSImpl) LoadJWKS4AuthToken(jwksPath string, interval time.Duration, reload bool) error {
+	jwks.filepath = jwksPath
+	if reload {
+		go func() {
+			for range time.Tick(interval) {
+				if err := jwks.load(); err != nil {
+					logutil.BgLogger().Error("Fail to load JWKS", zap.String("path", jwksPath), zap.Duration("interval", interval))
+				}
+			}
+		}()
+	}
+	return jwks.load()
+}
+
+// checkSigWithRetry verifies the signature in the jwt, and returns the claims.
+func (jwks *JWKSImpl) checkSigWithRetry(tokenString string, retryTime int) (map[string]interface{}, error) {
+	var (
+		verifiedPayload []byte
+		err             error
+	)
+	for retryTime >= 0 {
+		retryTime--
+		parts := strings.Split(tokenString, ".")
+		if len(parts) != 3 {
+			err = errors.New("Invalid JWT")
+			continue
+		}
+
+		// verify signature
+		verifiedPayload, err = jwks.verify(([]byte)(tokenString))
+		if err != nil {
+			if e := jwks.load(); e != nil {
+				err = errors.New("Fail to verify the signature and reload the JWKS")
```

Comment:
The error `e` is discarded so we won't know the actual reason.
We can just use `err = jwks.load()` because the message `Fail to verify the signature and reload the JWKS` is useless.